### PR TITLE
feat: send confirmation email

### LINKED ISSUE #4: feat: send confirmation email
When someone subscribe to the summary emails we should verify his email by sending him a confirmation email using a link with a code generated from an Ethereum signature.

### PR DESCRIPTION
**Pull request by wa0x6e**
Fixes #4 

This PR will edit the subscription workflow to:

- send a confirmation email on `subscribe`
- verify the confirmation when user click on the VERIFY link in the email

This feature requires an additional column in the database to distinguish users who have confirmed their email

> ALTER table subscribers ADD column verified BIGINT;

Email verification link will also include a unique signature (EIP712), to authenticate them 

This PR also add some new variables into .env

- `FRONT_HOST`: hostname of envelop-ui, used in templates for absolute link to envelop-ui
- `WALLET_X`: to hold the key for the local wallet. Only `WALLET_PRIVATE_KEY` is used by the code, other are for backups. Maybe we can remove them ?

---- 

## CORE TODO

- [x] Add `verify` API endpoint to receive the verify request
- [x] Send verification email with the verification link (waiting for PR #18 )
- [x] Add unique signature in the verification link

## EXTRA TODO

- [ ] Return a special error code when user is already subscribed
- [ ] Add dedicated endpoint to trigger sending the verification email again ? (in case first one is corrupted or not working for unknown reasons)

### CODE REVIEW

File: src/helpers/schema.sql [hunk 1]
```diff
@@ -2,6 +2,7 @@ CREATE TABLE subscribers (
   email VARCHAR(256) NOT NULL,
   address VARCHAR(256) NOT NULL,
   created BIGINT NOT NULL,
+  verified BIGINT,
```
Comment (bonustrack) [inline src/helpers/schema.sql:5]: This can be `INT`, and should not be `NULL`
```suggestion
  verified INT NOT NULL,
```

File: package.json [hunk 2]
```diff
@@ -24,6 +25,7 @@
     "cors": "^2.8.5",
     "cross-fetch": "^3.1.5",
     "dotenv": "^16.0.1",
+    "ethers": "5.6.2",
```
Comment (bonustrack) [inline package.json:28]: It's better to use just the dependency we need, I guess here is `@ethersproject/wallet`

File: .env.example [hunk 3]
```diff
@@ -1,3 +1,8 @@
 DATABASE_URL=
 SENDGRID_API_KEY=
 HOST=
+FRONT_HOST=
+
+WALLET_ADDRESS=
+WALLET_MNEMONIC=
```
Comment (bonustrack) [inline .env.example:7]: It's better remove these 2, we don't need a mnemonic and the public address can be generated from the private key.

File: .env.example [hunk 4]
```diff
@@ -1,3 +1,8 @@
 DATABASE_URL=
 SENDGRID_API_KEY=
 HOST=
+FRONT_HOST=
+
```
Comment (bonustrack) [inline .env.example:5]:
```suggestion
```

File: src/sign/index.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+const NAME = 'envelop';
+const VERSION = '1';
+
+const domain = {
+  name: NAME,
+  version: VERSION
+};
+
+const buildValue = (email: string, address: string, action: string): Subscription => {
```
Comment (bonustrack) [inline src/sign/index.ts:13]: Do we need this function?

File: src/sign/types.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  email: string;
+  action: string;
+}
+
+export const SubscriptionTypes = {
+  Subscriber: [
+    { name: 'address', type: 'address' },
+    { name: 'email', type: 'string' },
+    { name: 'action', type: 'string' }
+  ]
```
Comment (bonustrack) [inline src/sign/types.ts:12]: Would be better to have a message type for each actions like this:
```suggestion
  Subscribe: [
    { name: 'address', type: 'address' },
    { name: 'email', type: 'string' }
  ]
```

File: src/helpers/schema.sql (hunk 1 above)
Comment (wa0x6e) [reply src/helpers/schema.sql:5]: `verified` is similar to `created`, and recording the verification date.

It just give one more data, and fulfill the same purpose.

Should we downgrade to a simple 0|1 ?

Comment (bonustrack) [reply src/helpers/schema.sql:5]: Ok we can keep timestamp like created but i would remove NULL
```suggestion
  verified BIGINT NOT NULL,
```

**Comment by bonustrack** [comment]
Email doesn't include the signature, some promise is not resolved
image

File: src/sign/index.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+const buildValue = (email: string, address: string): Subscriber => {
+  return {
+    email,
+    address: getAddress(address)
+  };
+};
+
+const signer = new Wallet(process.env.WALLET_PRIVATE_KEY as string);
+
+const sign = async (message, type): Promise<string> => {
```
Comment (bonustrack) [inline src/sign/index.ts:22]: We don't need such function, it's only being used once

File: src/sign/index.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  return await signer._signTypedData(domain, type, message);
+};
+
+const verify = (message, signature: string, type): boolean => {
```
Comment (bonustrack) [inline src/sign/index.ts:26]: All the functions should use this format
```suggestion
function verify(message, signature: string, type): boolean {
```

**Comment by wa0x6e** [comment]
> Email doesn't include the signature, some promise is not resolved image

fixed

**Comment by bonustrack** [comment]
Looks good, was able to test the flow. Can you just make it so we can't confirm an email twice (so the verified field dont get updated again). Also on the UI we don't need a button "Verify email", we can verify directly when the user open the page and show success message.

**Comment by bonustrack** [comment]
`verified` wont be NULL it will be `0`, cuz I used `verified BIGINT NOT NULL,` on the db

**Pull request merged**